src/bridges/nft-basic/NFTVault.sol

```diff
@@ -0,0 +1,135 @@
+        AztecTypes.AztecAsset calldata,
+        uint256 _totalInputValue,
+        uint256 _interactionNonce,
+        uint64 _auxData,
+        address
+    )
+        external
+        payable
+        override(BridgeBase)
+        onlyRollup
+        returns (uint256 outputValueA, uint256 outputValueB, bool isAsync)
+    {
+        if (
+            _inputAssetA.assetType == AztecTypes.AztecAssetType.NOT_USED
+                || _inputAssetA.assetType == AztecTypes.AztecAssetType.ERC20
+        ) revert ErrorLib.InvalidInputA();
+        if (
+            _outputAssetA.assetType == AztecTypes.AztecAssetType.NOT_USED
+                || _outputAssetA.assetType == AztecTypes.AztecAssetType.ERC20
+        ) revert ErrorLib.InvalidOutputA();
+        if (_totalInputValue != 1) {
+            revert ErrorLib.InvalidInputAmount();
+        }
+        if (
+            _inputAssetA.assetType == AztecTypes.AztecAssetType.ETH
+                && _outputAssetA.assetType == AztecTypes.AztecAssetType.VIRTUAL
+        ) {
+            return (1, 0, false);
+        } else if (_inputAssetA.assetType == AztecTypes.AztecAssetType.VIRTUAL) {
+            NFTAsset memory token = nftAssets[_inputAssetA.id];
+            if (token.collection == address(0x0)) {
+                revert ErrorLib.InvalidInputA();
+            }
+
+            address to = REGISTRY.addresses(_auxData);
+            if (to == address(0x0)) {
+                revert ErrorLib.InvalidAuxData();
+            }
+            delete nftAssets[_inputAssetA.id];
+            emit NFTWithdraw(_inputAssetA.id, token.collection, token.tokenId);
+
+            if (_outputAssetA.assetType == AztecTypes.AztecAssetType.ETH) {
+                IERC721(token.collection).transferFrom(address(this), to, token.tokenId);
+                return (0, 0, false);
+            } else {
+                IERC721(token.collection).approve(to, token.tokenId);
+                NFTVault(to).matchAndPull(_interactionNonce, token.collection, token.tokenId);
+                return (1, 0, false);
+            }
+        }
+    }
+
+    /**
+     * @notice Function for the second step of a NFT deposit or for transfers from other NFTVaults.
+     * @dev For a deposit, this method is called by an Ethereum L1 account that owns the NFT to deposit.
+     * The user must approve this bridge contract to transfer the users NFT before this function
+     * is called. This function assumes the NFT contract complies with the ERC721 standard.
+     * For a transfer from another NFTVault, this method is called by the NFTVault that is sending the NFT.
+     *
+     * @param _virtualAssetId - the virutal asset id of the note returned in the deposit step of the convert function
+     * @param _collection - collection address of the NFT
+     * @param _tokenId - the token id of the NFT
+     */
+
+    function matchAndPull(uint256 _virtualAssetId, address _collection, uint256 _tokenId) external {
```

Comment:
@LHerskind, @benesjan pointed out a griefing attack vector here. someone can register any NFT with any `virtualAssetId`. So if you get a virtual asset id with the intention of registering your NFT, i could just register any NFT in its place, so you can't deposit your NFT--that virtual asset id has been taken. There may be incentive to do this if we build a bridge for NFTs w/ governance rights, or something similar.

For a solution, we could have a depositor specify the collection address of the NFT that they want to deposit in step 1 of the deposit flow--they would pass the `id` (from the AddressRegistry) of the address of the collection in `auxData`, and during the deposit step the `convert` function would update the collection address in the corresponding `nftAssets` mapping. The `NFTAsset` struct would also include a `isMatchPending` flag, which would be set to true during the first step of the deposit flow as well.

Then in `matchAndPull`, there would be a check to make sure `nftAssets[_virtualAssetId].collection == _collection` and `isMatchPending == true`. Then the `nftAssets` mapping entry would be updated with the tokenId and set `isMatchPending` to false. This way the only "attack" would be for you to deposit an NFT from the same collection to my aztec account, costing me nothing.

Reply:
Hmm, ye, good point. It makes the onboarding a bit more annoying, but ye for high-value cases you could be pretty annoying by frontrunning every time someone tries to `matchAndPull` 🤖. It will probably still be a vector for some of the NFTs where many "collections" live under the same address from the same brand. Can't remember the name, but believe one of the larger ones did so, artblocks if I recall correctly. But still much more limited. 

The other attack still cost you something as you would lose gas money, but that should be small in comparison to the amount to grieve you for any non-valueless NFT.

Reply:
Alternatives

1. we could require that the collection and the tokenId are passed in step 1 of the deposit, so it's not ambiguous which token the virtual asset is for. Then we wouldn't need the `isMatchPending` flag, since the virtual asset is only good for 1 token. 

2. we could add another field to `NFTAsset` that specifies an `address` that can deposit NFTs to that virtual asset.

I think 1 might be the best solution

Reply:
There is no `isMatchPending` in the NFTVault contract.
 
If you are using the `auxData` you would be able to fit most thing in, but some `tokenId`s would not fit. 

I think going with sol 1 is probably a good way to handle it yes. If you update `NFTAsset` to be a struct of 3 elements, you can write the collection and `tokenId` directly and then it is just the flag for it being "deposited" that needs to be updated when matched? So storage wise should not be too different in costs. If we are anyway using the `auxData` such that not all can fit in, these 3 elements can also fit into one slot, so gas costs gets a bit lower with the restriction that addressable NFT space. 
```solidity
struct NFTAsset {
  address collection;
  uint48 tokenId;
  bool isInEscrow;
}
```

Reply:
> There is no isMatchPending in the NFTVault contract.

Right, I was just building off of the solution suggested in the first comment.

If we have users specify the collection and `tokenId`, do you think we need the `isInEscrow` flag?

With the restriction of specifying all of the token info in the first step of a deposit, a virtual asset id will always only be associated with 1 NFT. You'd be able to tell if the NFT was in escrow by looking up the owner of the NFT on the NFT contract. It might be a bit more convenient to do it with the `isInEscrow` flag in the NFTVault, but not sure if it is worth it.

Reply:
also just thinking that if we pass the current NFT owner, collection and tokenId into the deposit flow, we can deposit the NFT into aztec in 1 transaction.

The owner will need to approve the bridge to do a `transferFrom` beforehand, but they could pass the owner to `transferFrom` as an id from the address registry as well. 

This would be a better UX, but would be more restrictive in terms of input options.

`owner` and `collection` could each be 20 bits and offer ~1 million options each. 
`tokenId` could be 24 bits and offer ~17 million options.

Do you foresee any problems with this approach? I think the UX improvements would be worth the input limitations

Reply:
If the owner approves the NFT beforehand. Anyone can claim it. The contract should only ever pull NFTs or tokens from `msg.sender`, otherwise, you could use the rollup to steal the approved NFTs where the thief even will be hidden while doing it.